### PR TITLE
[FLINK-13363][docs] Add documentation for streaming aggregate performance tuning

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

Add documentation for streaming aggregate performance tuning

## Brief change log

- Add introduction for streaming aggregate performance tuning in `dev/table/tuning/streaming_aggregation_optimization.md`.

## Verifying this change
only add docs without test

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

File: docs/dev/table/tuning/streaming_aggregation_optimization.md [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+under the License.
+-->
+
+SQL is the most widely used language for data analytics. Flink's Table API and SQL enables users to define efficient stream analytics applications in less time and effort. Moreover, Flink Table API and SQL is effectively optimized, it integrates a lot of query optimizations and tuned operator implementations. But not all of the optimizations are enabled by default, so for some workloads, it is possible to improve performance by turning on some options.
+
+In this page, we will introduce some useful optimization options and the internals of streaming aggregation which will bring great improvement in some cases.
+
+<span class="label label-danger">Attention</span> Currently, the optimization options mentioned in this page are only supported for the Blink planner.
+
+<span class="label label-danger">Attention</span> Currently, the streaming aggregation optimization are only supported for [unbounded-aggregations]({{ site.baseurl }}/dev/table/sql.html#aggregations). Optimizations for [window aggregations]({{ site.baseurl }}/dev/table/sql.html#group-windows) will be supported in the future.
```
Comment (leonardBang) [inline docs/dev/table/tuning/streaming_aggregation_optimization.md]: "the streaming aggregation **optimization** are"  should be "the streaming aggregation **optimizations** are"

File: docs/dev/table/tuning/streaming_aggregation_optimization.md [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+The following figure explains how the mini-batch aggregation reduces state operations.
+
+<div style="text-align: center">
+  <img src="{{ site.baseurl }}/fig/table-streaming/minibatch_agg.png" width="50%" height="50%" />
+</div>
+
+MiniBatch optimization is disabled by default. In order to enable this optimization, you should set options `table.exec.mini-batch.enabled`, `table.exec.mini-batch.allow-latency` and `table.exec.mini-batch.size`. Please see [configuration]({{ site.baseurl }}/dev/table/config.html#execution-options) page for more details.
+
+The following examples show how to enable the option.
```
Comment (leonardBang) [inline docs/dev/table/tuning/streaming_aggregation_optimization.md]: Maybe  "enable the **optimization**"  or “set these **options**”  would be better

File: docs/dev/table/tuning/streaming_aggregation_optimization.md [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+<div data-lang="java" markdown="1">
+{% highlight java %}
+// instantiate table environment
+TableEnvironment tEnv = ...
+
+tEnv.getConfig()        // access high-level configuration
+  .getConfiguration()   // set low-level key-value options
+  .setString("table.exec.mini-batch.enabled", "true")  // enable mini-batch optimization
+  .setString("table.exec.mini-batch.allow-latency", "5 s") // use 5 seconds to buffer input records
+  .setString("table.exec.mini-batch.size", "5000"); // the maximum number of records can be buffered by each aggregate operator tasks
```
Comment (leonardBang) [inline docs/dev/table/tuning/streaming_aggregation_optimization.md]: "each aggregate operator **task**" or "each **of** aggregate operator tasks"  would be better

File: docs/dev/table/tuning/streaming_aggregation_optimization.md [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+// instantiate table environment
+val tEnv: TableEnvironment = ...
+
+tEnv.getConfig         // access high-level configuration
+  .getConfiguration    // set low-level key-value options
+  .setString("table.exec.mini-batch.enabled", "true") // enable mini-batch optimization
+  .setString("table.exec.mini-batch.allow-latency", "5 s") // use 5 seconds to buffer input records
+  .setString("table.exec.mini-batch.size", "5000") // the maximum number of records can be buffered by each aggregate operator tasks
```
Comment (leonardBang) [inline docs/dev/table/tuning/streaming_aggregation_optimization.md]: "each aggregate operator **task**" or "each **of** aggregate operator tasks" would be better

File: docs/dev/table/tuning/streaming_aggregation_optimization.md [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+<div data-lang="python" markdown="1">
+{% highlight python %}
+# instantiate table environment
+t_env = ...
+
+t_env.get_config()        # access high-level configuration
+  .get_configuration()    # set low-level key-value options
+  .set_string("table.exec.mini-batch.enabled", "true") # enable mini-batch optimization
+  .set_string("table.exec.mini-batch.allow-latency", "5 s") # use 5 seconds to buffer input records
+  .set_string("table.exec.mini-batch.size", "5000"); # the maximum number of records can be buffered by each aggregate operator tasks
```
Comment (leonardBang) [inline docs/dev/table/tuning/streaming_aggregation_optimization.md]: "each aggregate operator **task**" or "each **of** aggregate operator tasks" would be better

File: docs/dev/table/tuning/streaming_aggregation_optimization.md [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+{% endhighlight %}
+</div>
+</div>
+
+## Local-Global Aggregation
+
+Local-Global is proposed to solve data skew problem by dividing an group aggregation into two stages, that is doing local aggregation in upstream firstly, and followed by global aggregation in downstream, which is similar to Combine + Reduce pattern in MapReduce. For example, considering the following SQL:
```
Comment (leonardBang) [inline docs/dev/table/tuning/streaming_aggregation_optimization.md]: "dividing **an** group aggregation"->"dividing **a** group aggregation"

File: docs/dev/table/tuning/streaming_aggregation_optimization.md [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+
+{% highlight sql %}
+SELECT color, sum(id)
+FROM T
+GROUP BY color
+{% endhighlight %}
+
+It is possible that the records in the data stream are skewed, thus some instances of aggregation operator have to process much more records than others, which leads to hotspot.
+The local aggregation can help to accumulate a certain amount of inputs which have the same key into a single accumulator. The global aggregate will only receive the reduced accumulators instead of large number of raw inputs.
```
Comment (leonardBang) [inline docs/dev/table/tuning/streaming_aggregation_optimization.md]: "The global **aggregate** will" -> "The global **aggregation** will".
keep same as mentioned above.

File: docs/dev/table/tuning/streaming_aggregation_optimization.md [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+ day,
+ COUNT(DISTINCT user_id) AS total_uv,
+ COUNT(DISTINCT user_id) FILTER (WHERE flag IN ('android', 'iphone')) AS app_uv,
+ COUNT(DISTINCT user_id) FILTER (WHERE flag IN ('wap', 'other')) AS web_uv
+FROM T
+GROUP BY day
+{% endhighlight %}
+
+Flink SQL optimizer can recognize the different filter arguments on the same distinct key. For example, in the above example, all the three COUNT DISTINCT are on `user_id` column.
+Then Flink can use just one share state instance instead of three state instances to reduce state access and state size. In some workloads, this can get significant performance improvements.
```
Comment (leonardBang) [inline docs/dev/table/tuning/streaming_aggregation_optimization.md]: replacing “share” with "shared" here would be better?
“just one **share** state instance”->“just one **shared** state instance”

**Comment by wuchong** [comment]
Thanks @leonardBang for the careful reviewing. I addressed the comments.

**Comment by wuchong** [comment]
Thanks @leonardBang , I will merge this then.

**Pull request closed without merging**